### DISCUSSION TITLE
INA226 voltage scale

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#16643] [Q&A]
Moving discussion from Discord to Github.

Background:

INA226 DC current meter can measure voltages up to 36v. For higher voltage DC systems, PZEM-017 is a possibility, but also a simple voltage divider can reduce voltage to useable range for INA226. However, driver needs to be updated to deal with the correct multiplier so that actual voltage is displayed, and correct watts are reported.

@barbudor is adjusting [code](https://github.com/barbudor/Tasmota/tree/ina226_sensor54_voltage_fs) and I am testing.

Currently, code is crashing upon entering the command `sensor54 13 81.92`

**Comment by barbudor** [comment]
looking into it now @pkkrusty

**Reply by barbudor** [reply]
I can't really reproduce the crash but there was an AddLog("%_f") that was not working as expected. So I removed it.
I'm on ESP8266, are you on ESP8266 or ESP32 ?

I've pushed an update to my repo. Can you try it ?

```
14:12:36.051 CMD: status 8
14:12:36.066 MQT: stat/dev-4119/STATUS8 = {"StatusSNS":{"Time":"2022-09-24T14:12:36+02:00","INA226-1":{"Id":0,"Voltage":4.78,"Current":0.000,"Power":0}}}
14:12:43.504 CMD: sensor54 10
14:12:43.518 MQT: stat/dev-4119/SENSOR = {"Sensor54-device-settings-1":{"SHUNT_R":0.100000,"FS_I":2.0"FS_V":40.96}}
14:12:51.587 CMD: sensor54 13 81.92
14:12:51.597 MQT: stat/dev-4119/SENSOR = {"Sensor54-device-settings-1":{"SHUNT_R":0.100000,"FS_I":2.0"FS_V":81.92}}
14:13:10.984 CMD: status 8
14:13:11.001 MQT: stat/dev-4119/STATUS8 = {"StatusSNS":{"Time":"2022-09-24T14:13:10+02:00","INA226-1":{"Id":0,"Voltage":9.56,"Current":0.000,"Power":0}}}
```

**Reply by pkkrusty** [reply]
ESP32, but will also test on 8266.

**Reply by barbudor** [reply]
Different esp may explain why it displays 'invalid' in my case and crash in your casd
Nevertheless the latest ino should be fine.

**Reply by pkkrusty** [reply]
Yep, that fixed it.

**Reply by barbudor** [reply]
Great, I will PR later today

**Reply by pkkrusty** [reply]
New problem. Power calc is not using new voltage value, but using original (in this case ~18v)
![IMG_4213](https://user-images.githubusercontent.com/79770016/192140884-d9947faf-92dd-41b3-ba54-b5ac85d54565.PNG)

**Reply by barbudor** [reply]
OK, let me check

**Reply by barbudor** [reply]
Fixed.
The power was read from a chip register which was not aware of the new voltage. So now it is calculated from V * I

**Reply by pkkrusty** [reply]
Will work on updating documentation unless you want to.

**Reply by barbudor** [reply]
No problem, I leave the honour to you 😄 

PR: https://github.com/arendst/Tasmota/pull/16653

**Reply by Aint-Nobody** [reply]
@pkkrusky

I try to get a ina226 running, but no sucsess.
Can you help?

https://github.com/arendst/Tasmota/discussions/24673#discussion-9935541

**Comment by pkkrusty** [comment]
Somewhat related issue @barbudor : Power values are always positive, even when current is negative (like in my solar setup where INA226 is measuring current into and out of the battery). 
![IMG_4265](https://user-images.githubusercontent.com/79770016/194263884-ad155854-890c-4ad0-b00d-2537e30bfe68.PNG)

**Reply by barbudor** [reply]
OK. I will look tonight

**Reply by barbudor** [reply]
no idea so far,
voltage, current and power are all 3 "floats"
and power = voltage * current
So if current is negative, power should be too

Are you compiling yourself ?
Could you test with removing the `//` comment in front of the lines 320..323 here:
https://github.com/arendst/Tasmota/blob/7f0852dca549d9be283560cd4463c47030b8f45c/tasmota/tasmota_xsns_sensor/xsns_54_ina226.ino#L314-L324

With `WebLog 3` you should see exactly what is measured and calculated in real-time
Thanks

If you can't, I'll setup a test tomorrow evening

**Reply by pkkrusty** [reply]
I'll try it out, but may take a few days until I have some free time.

**Reply by pkkrusty** [reply]
I wonder if it's a display issue, I.e. the web GUI displays X number of significant digits and the negative symbol gets truncated somewhere along the line. Because you're right, float numbers multiplied should give correct negative values. 

My brain is tickling me about sometimes I feel like I see negative values, and other times not. Let me pull my influxdb data and see if I can find a pattern.

**Reply by pkkrusty** [reply]
<img width="720" alt="Screen Shot 2022-10-06 at 21 43 03" src="https://user-images.githubusercontent.com/79770016/194424087-408460c2-fa7f-4719-8b2b-45c880219733.png">
Nah, current goes negative but power is always positive. I'll look at some of my other devices tomorrow.

**Reply by joba-1** [reply]
I think ina226 power is always positive „by design“

**Reply by joba-1** [reply]
… and the datasheet confirms that the MSB for the current register is the sign bit and for the power register it is „normal“ data.
So sign for power needs to be derived from voltage and current sign bits (xor)

**Reply by barbudor** [reply]
@pkkrusty I can't reproduce : negative current => negative power
Are you sure you are using the latest version ?
![image](https://user-images.githubusercontent.com/2996042/194641886-d884eaff-9f8b-487e-8c95-43511985bd79.png)

@joba-1 : not using the power register, just P = V * I

**Reply by joba-1** [reply]
ok then, if it were my issue I'd bet on me having checked out the wrong branch while pulling your changes :)

**Comment by norbert-walter** [comment]
@pkkrusty 
May I ask how you implemented the BMS-2 in Tasmota? I see they use an INA226 and read the cell voltages. How do you measure the cell voltages? With the ESP32 or an external module? Can you explain their structure to me? I would like to do something similar too.

**Reply by pkkrusty** [reply]
I'm using a BMS module from a Mercedes electric vehicle that uses the LTC6802-1 chip. I communicate with that using scripting via the SPI bus. Pretty specific to the hardware I had available.

**Reply by pkkrusty** [reply]
Measuring cell voltages is pretty tricky if you try to DIY. Lot's of things to consider.

**Comment by norbert-walter** [comment]
Ok I understand. You using scripting internal in tasmota or external?

**Comment by norbert-walter** [comment]
Can you me send your scrip code? I want to look the implementation.

**Reply by pkkrusty** [reply]
Sure. I started a discussion thread a few months ago with the setup and the script. Search for my username.

**Reply by pkkrusty** [reply]
#16257

**Comment by tasmotari** [comment]
I have INA226 with R002 shunt,
but I get incorrect voltage:

13.06V battery = 5.22V
1.26V battery = 0.5V

It looks like need 2.5 multiplier

How to fix it?

**Reply by barbudor** [reply]
Cross posting will not help and just annoy us
I have answered to your 1st post
